client/network/src/service/out_events.rs

```diff
@@ -0,0 +1,236 @@
+pub fn channel() -> (Sender, Receiver) {
+	let (tx, rx) = mpsc::unbounded();
+	let stats = Arc::new(Mutex::new(None));
+	let tx = Sender { inner: tx, stats: stats.clone() };
+	let rx = Receiver { inner: rx, stats };
+	(tx, rx)
+}
+
+/// Sending side of a channel.
+///
+/// Must be associated with an [`OutChannels`] before anything can be sent on it
+///
+/// > **Note**: Contrary to regular channels, this `Sender` is purposefully designed to not
+/// > implement the `Clone` trait. If someone adds a `#[derive(Clone)]` below, it is **wrong**.
+pub struct Sender {
+	inner: mpsc::UnboundedSender<Event>,
+	/// Clone of [`Receiver::stats`].
+	stats: Arc<Mutex<Option<Arc<Mutex<StatsCollect>>>>>,
+}
+
+impl fmt::Debug for Sender {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Sender").finish()
+	}
+}
+
+/// Receiving side of a channel.
+pub struct Receiver {
+	inner: mpsc::UnboundedReceiver<Event>,
+	/// Initially contains `None`, and will be set to a value once the corresponding [`Sender`]
+	/// is assigned to an instance of [`OutChannels`].
+	stats: Arc<Mutex<Option<Arc<Mutex<StatsCollect>>>>>,
+}
+
+impl Stream for Receiver {
+	type Item = Event;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Event>> {
+		if let Some(ev) = ready!(Pin::new(&mut self.inner).poll_next(cx)) {
+			let stats = self.stats.lock();
+			if let Some(stats) = stats.as_ref() {
+				decrease_stats(&mut *stats.lock(), &ev);
+			} else {
+				log::warn!("Inconsistency in out_events: event happened before sender associated");
+			}
+			Poll::Ready(Some(ev))
+		} else {
+			Poll::Ready(None)
+		}
+	}
+}
+
+impl fmt::Debug for Receiver {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Receiver").finish()
+	}
+}
+
+impl Drop for Receiver {
+	fn drop(&mut self) {
+		// Empty the list to properly decrease the stats.
+		while let Some(Some(_)) = self.next().now_or_never() {}
+	}
+}
+
+/// Statistics about the content of the channel.
+#[non_exhaustive]
+pub struct StatsCollect {
+	/// For each protocol, total number of substream open notifications in the channel.
+	pub notifications_open_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, total number of substream closed notifications in the channel.
+	pub notifications_closed_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, the total size of all notifications stuck in the channel.
+	pub notifications_sizes_total: HashMap<ConsensusEngineId, u64>,
+
+	/// Number of DHT events in the channel.
+	///
+	/// > **Note**: DHT events are extremely rare at the moment, but this could be broken down
+	/// >			further if there's a need to.
+	pub dht_events: u64,
+}
+
+/// Collection of senders.
+pub struct OutChannels {
+	event_streams: Vec<Sender>,
+	/// The stats we collect. A clone of this is sent to each [`Receiver`] associated with this
+	/// object.
+	stats: Arc<Mutex<StatsCollect>>,
```

Comment:
Do I understand correctly that all communication from the network to any upper component needs to lock this lock (each time a message is send into a channel and each time one is taken out of a channel)? If so, I would expect this to introduce a lot of contention.